nixos/modules/config/ldso.nix

```diff
@@ -0,0 +1,61 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) last splitString mkOption types mdDoc mkMerge escapeShellArg mkIf;
+
+  libDir = pkgs.stdenv.hostPlatform.libDir;
+  ldsoBasename = last (splitString "/" pkgs.stdenv.cc.bintools.dynamicLinker);
+
+  pkgs32 = pkgs.pkgsi686Linux;
+  libDir32 = pkgs32.stdenv.hostPlatform.libDir;
+  ldsoBasename32 = last (splitString "/" pkgs32.stdenv.cc.bintools.dynamicLinker);
+in {
+  options = {
+    environment.ldso = mkOption {
```

Comment:
This is a manner of taste and a bit of an architectural refactor but I'd find it more sensible if the different ldsos available in NixOS were all encompassed by one option.

Either through sub-options for each (`environment.ldso.stub`, `environment.ldso.nix-ld` etc.) or, since there can only ever be one ldso, an interface like this:

```
config.environment.ldso.package = pkgs: pkgs.stub-ld; # default
```

```
config.environment.ldso.package = pkgs: pkgs.nix-ld; # customisation
```

```
config.environment.ldso.enable = false; # disable
```

(You could be even more minimalist about it using only the package option (setting it null to disable) but that doesn't feel abstract enough to me.)

An interface like this would make everything a lot simpler as removes the need to have separate modules for the different ldsos, only packages (that's really what they are tbh).  
This in turn also makes it much simpler to add other ldsos in NixOS in the future as all you'd have to do is create a new package and users can simply plug it into the `package`.

Reply:
> . I could make `stub-ld` a package, but then the message would be set in very much the wrong place, imo.

I can see where you're coming from but, OTOH it's also a binary artifact built from some source code. That's a package to me, just like nix-ld.

> Also, users would need to set 2 options to disable `stub-ld` instead of one (one for the main arch and one for 32bit).

Why? As mentioned, all it'd take to disable in my design would be

```
config.environment.ldso.enable = false; # disable
```

Additionally, overriding a package is also simplified because you only need to set one option.

> I also need to consider backward compatibility for `nix-ld` users, so that pretty much brings us back to the current implementation.

Indeed looking at the nix-ld module, it does a bit more than setting the package.

I still find it simpler to have this option this way such that you don't have to do the `environment.stub-ld.install = false;` dance in every ldso module.

Reply:
> Why? As mentioned, all it'd take to disable in my design would be
> 
> ```
> config.environment.ldso.enable = false; # disable
> ```

How would this work, though? keep in mind that `nix-ld` only replaces the elf loader for the main arch, and we still need to use the stub for 32bit even when `nix-ld` is installed.

Reply:
> I still find it simpler to have this option this way such that you don't have to do the `environment.stub-ld.install = false;` dance in every ldso module.

That much can probably be done more reasonably just by having the `stub-ld` module set `environment.ldso{,32}` with `mkDefault`. Then `nix-ld` doesn't need to do any extra steps to override it, and neither do users.